crates/ruff_python_parser/src/lexer.rs

```diff
@@ -579,6 +718,19 @@ impl<'source> Lexer<'source> {
     // This is the main entry point. Call this function to retrieve the next token.
     // This function is used by the iterator implementation.
     pub fn next_token(&mut self) -> LexResult {
+        if let Some(fstring_context) = self.fstring_stack.last() {
+            if !fstring_context.is_in_expression() {
+                self.cursor.start_token();
+                if let Some(tok) = self.maybe_lex_fstring_middle()? {
+                    return Ok((tok, self.token_range()));
+                }
+                if let Some(tok) = self.maybe_lex_fstring_end() {
+                    self.fstring_stack.pop();
+                    return Ok((tok, self.token_range()));
+                }
```

Comment:
For which tokens does fstring_middle` and `fstring_end` both return None? Could we instead branch early on these tokens and unconditionally either return the middle or end part?

Reply:
So, `maybe_lex_fstring_middle` will return `None` when:
1. The first character is `{` not followed by another `{` which is the start of f-string expression. For example, `f"{foo}"` and not `f"{{foo}}"`
2. Empty f-string which means that we should directly consume the f-string end. For example, `f""`, `f''`, `f""""""`, `f''''''`.

I think you're referring to (1) then.

Reply:
This is a really good suggestion. Thanks, I've added it here [`5b41241` (#6659)](https://github.com/astral-sh/ruff/pull/6659/commits/5b41241b0f37d7539439c93a3408d91f69c29861) with some test cases.

Reply:
I've added `unreachable` to check in what other cases this happens.

Reply:
There's another case as well:

We're in format specifier mode and we've reached the end i.e., the next character is `}` which should be emitted as RBrace.

For example, in `f"{foo:.3f}"`, the format specifier `.3f` is emitted as `FStringMiddle`, then in the next iteration we need to emit `}` (RBrace) but we'll try to lex f-string middle (returns `None`) and we're not at the end of f-string.

Now, I'm unable to find any other case which reaches `unreachable`.

Reply:
I do like this idea but it sacrifices readability a bit (https://github.com/astral-sh/ruff/pull/6659#discussion_r1305569804). I'll give it a bit of thought but I'm mostly leaning towards reverting the change.